Added setters for log enablement, all-log enablement, and log level.

cathalgarvey opened the pull request:
To make this make sense, also had to add code that checks whether
logs are enabled at call time rather than checking the value of
`IsEnabled` which is set at instantiation time. The value of IsEnabled
is updated whenever an internal method that uses it checks for
enablement.

Added a little bit of documentation to existing and new functions.

cathalgarvey commented:
Forgot to reference #3 :)

azer commented on settings.go:
```diff
@@ -22,35 +33,70 @@ func Enabled() (map[string]bool, bool) {
 	for _, key := range keys {
-		all[key] = true
+		// Trim in case people pass space-gapped names.
+		all[strings.TrimSpace(key)] = true
```
good idea :+1:

azer commented on logger.go:
```diff
@@ -18,12 +18,22 @@ func New(name string) *Logger {
+// This is a little kludgy but was introduced to allow dynamic log-name/log-level
+// changes to be made by programs at runtime. IsEnabled property is retained for API
+// compatibility.
+func (l *Logger) amIEnabled() bool {
```
this naming confuses me. how is it different with `IsEnabled` ? Why do we have both ?

azer commented:
thanks a lot, I'm leaving some comments

azer commented on settings.go:
```diff
@@ -1,17 +1,28 @@
+// Enabled - Returns the map representing enabled logs, and whether all loggers are
+// enabled regardless of name. These are initialised from environment variables
+// at start-up.
 func Enabled() (map[string]bool, bool) {
```
I think we can remove this, and change `enabled` `allEnabled` to capital case

azer commented on settings.go:
```diff
@@ -1,17 +1,28 @@
+// initEnabled - Gets the LOG environment variable, splits on commas, and stores the
+// values as loggers that are permitted to print. Does not validate that loggers
+// exist; typos may break loggers silently.
+func initEnabled() (map[string]bool, bool) {
```
how about making this just `init` and setting `enabled` and `allEnabled` from inside of `init` ?

azer commented on settings.go:
```diff
@@ -22,35 +33,70 @@ func Enabled() (map[string]bool, bool) {
 
-	if strings.ToUpper(level) == "ERROR" {
+// Populates default value of 'verbosity' variable from Envvar.
+func initVerbosity() int {
```
Can we also do this in the `init` function ?

azer commented on settings.go:
```diff
@@ -22,35 +33,70 @@ func Enabled() (map[string]bool, bool) {
-	return 1
+// Verbosity - Get the value of the LOG_LEVEL environment variable, converted to an integer.
+// Default/Invalid value is 1. If "TIMER", this is 2. If "ERROR", this is 3.
+func Verbosity() int {
```
I think we can remove this, too, by just making `verbosity` variable capital case (public)

cathalgarvey commented:
By all means; all of the above oddities derived from me avoiding changing the existing API. If you feel like changing APIs there are much cleaner ways to expose the required functionality without kludges.

For e.g.: `Logger.IsEnabled` is a boolean field, and I assumed it should be kept for client programmers to rely upon, so I created a hidden method to take over for internal code, and update IsEnabled as a side effect to help client code.

If I were breaking the API, I'd just drop the `IsEnabled` field and replace it with a method that does what `amIEnabled` does.

Likewise the `Enabled()` function could be removed if the `enabled` map were public, and we could dispense with getters and setters entirely. :)

azer commented on settings.go:
```diff
@@ -22,35 +33,70 @@ func Enabled() (map[string]bool, bool) {
+// * 1 - Info, Timer and Error
+// * 2 - Timer and Error
+// * 3 - Error only
+func SetVerbosity(level int) error {
```
It'd be nicer to take advantage of constants; 

```
const (
  TIMER_LEVEL = 1
  ERROR_LEVEL = 2
  INFO_LEVEL = 3
)
```

azer commented:
@cathalgarvey I think we can change the API little bit. This PR will enable logger to be used programmatically for first time, so it's ok change things for good looking API. Thanks a lot!

cathalgarvey commented:
Looking over Logger's importers on Godoc: https://godoc.org/github.com/azer/logger?importers

..Looks like nobody is using the sections of the API that I'm changing here. Well, maybe you are but I'm not looking at your other projects, because it's your API to break. :)

So we're safe.

cathalgarvey commented:
I have *not yet tested this*, I'm just pushing while in-progress.

Big changes:

* Exposed `Enabled`, `AllEnabled`, `Verbosity`.
* `Verbosity` is now a custom `int` type with three constant levels provided, private type.
* Logger now has a new method, `IsEnabled() bool`, to replace the old attribute. It doesn't look like any clients of this library were using or relying on the attribute, at least according to godoc. I removed the global `IsEnabled` function.
* `New(name string) *Logger` now adds `name` to the `Enabled` map if not already present, with default value of `false`. This exposes a minor race condition, but to resolve this would require hiding `Enabled` and using getters, setters, and `sync.Mutex`; because this is a logging package I figured the risk is low, but given the JSON portion of the API it's possible this would be used for (important) Telemetry, so maybe mutexed get/sets for Enabled are the best way to go? With your permission I'd rather hide `Enabled` again and mutex it with a getter/setter.
* Small changes where I thought they made sense; eg:
 removed what looked like a redundant `if` in PrettyPrefix and an overkill `fmt` call.

Other:

* I added some comments to document stuff
* I added some thoughts to the JSON encoding bits, because it looks right now like they may be vulnerable to invalid data, and because I think they could be more maintainable as a call to `json.Marshal`, possibly with the addition of a private struct to handle the intermediate representation or using an anonymous struct.
* My metalinter plugin auto-changed some tiny formatting stuff (sorry)

Next steps:

cathalgarvey commented:
Derp, typo posted early.

Next steps I'd recommend:

* Make choice between pretty printing and JSON an explicit developer choice, with the default choice being "decide dynamically" (e.g. as things are right now) and other choices being to either always JSON or always prettify the output.
* Make some tests..
* Change the JSON handling code to rely on `json.Marshal`

azer commented on format.go:
```diff
@@ -33,10 +31,26 @@ func (l *Logger) Format(verbosity int, sort string, msg string, attrs *Attrs) st
+	// j, err := json.Marshal(jstruct)
+	// if err != nil { return err }
+	// return string(j)
 	return fmt.Sprintf("{ \"time\":\"%s\", \"package\":\"%s\", \"level\":\"%s\",%s \"msg\":\"%s\" }", time.Now(), l.Name, sort, attrs, msg)
```
Definitely agree with you :+1: anonymous struct would be much better.

azer commented:
sorry for late merge, thanks for the PR :+1:

azer commented:
@cathalgarvey I can add you as collaborator if you'd like to merge your PRs quicker
